### PR TITLE
[#10146] Refactor get session result logic in the backend

### PR DESCRIPTION
**Pull request by xpdavid**
Part of #10146 

**Outline of Solution**

First commit:
- Correct one type definition in the frontend. Basically, `recipientEmail` can be null for anonymous case.

Second commit:
- Fix data inconsistency in the `typicalDataBundle`. The `giverSection` and `recipientSection` should follow those of its corresponding response.

Third commit:
- I purely add a common method for both student and instructor to get feedback session result. I also create a new bundle called `SessionResultsBundle` (The old `FeedbackSessionResultsBundle` is too complicated to change)
    - All responses and comments are fetched first and them checked one by one (so it is a compute-intensive operation). The old method fetches viewable responses by issuing multiple queries (communication-intensive operation)
    - Remove all complicated and unnecessary maps in the bundle and purely use `CourseRoster` to get name, teamName and sectionName.
    - Move the giver/recipient name hiding logic to `SessionResultData`. Similarly, data transformation will all be made there. The hiding logic will also not modify fields in `*Attributes`.
    - I have gone through the code base and look for session result related tests. The tests are migrated to test the new method as much as possible.
    - Rewrite the building process of `SessionResultData`. Also applied builder pattern to make the constructor more readable.

We decide to move irrelevant code in a separate PR as it related to the functionality of the CSV generation.

<strike>First commit:
- I purely add a common method for both student and instructor to get feedback session result. I also create a new bundle called `SessionResultsBundle` (The old `FeedbackSessionResultsBundle` is too complicated to change)
    - All responses and comments are fetched first and them checked one by one (so it is a compute-intensive operation). The old method fetches viewable responses by issuing multiple queries (communication-intensive operation)
    - Remove all complicated and unnecessary maps in the bundle and purely use `CourseRoster` to get name, teamName and sectionName.
    - Move the giver/recipient name hiding logic to `SessionResultData`. Similarly, data transformation will all be made there. The hiding logic will also not modify fields in `*Attributes`.
    - I have gone through the code base and look for session result related tests. The tests are migrated to test the new method as much as possible.

Second commit:
- Remove all unused methods related to session result.
- **Remove** get CSV functionality. I believe we can generate the CSV in the frontend by using the RESTFul data returned. Anyway, I need to remove them or the dependencies can be de-coupled.

Third commit:
- Just move a method to the correct place.</strike>

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/CourseRoster.java [hunk 1]
```diff
@@ -15,12 +16,24 @@
...
+    private final Map<String, StudentAttributes> studentListByEmail = new HashMap<>();
+    private final Map<String, InstructorAttributes> instructorListByEmail = new HashMap<>();
+    private final Map<String, List<StudentAttributes>> teamToMembersTable = new HashMap<>();
 
     public CourseRoster(List<StudentAttributes> students, List<InstructorAttributes> instructors) {
         populateStudentListByEmail(students);
         populateInstructorListByEmail(instructors);
+
+        // group students by team
+        for (StudentAttributes studentAttributes : this.getStudents()) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/CourseRoster.java:33]: Apply SLAP principle here: extract this block to a private method, no matter how simple it is.

File: src/main/java/teammates/common/datatransfer/CourseRoster.java [hunk 2]
```diff
@@ -104,4 +128,42 @@ private void populateInstructorListByEmail(List<InstructorAttributes> instructor
...
+            sectionName = Const.DEFAULT_SECTION;
+        } else if (isTeam) {
+            StudentAttributes teamMember = this.getTeamToMembersTable().get(identifier).iterator().next();
+
+            name = identifier;
+            teamName = identifier;
+            sectionName = teamMember.getSection();
+        }
+
+        return new String[] { name, teamName, sectionName };
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/CourseRoster.java:167]: While this works, we can just define another small data structure with 3 fields for this. That way we also don't need the `INFO_*` variables.

File: src/main/java/teammates/logic/api/Logic.java [hunk 3]
```diff
@@ -1678,83 +1434,20 @@ public FeedbackSessionResultsBundle getFeedbackSessionResultsForInstructorToSect
...
+     *
+     * @see FeedbackSessionsLogic#getSessionResultsForUser(String, String, String, UserRole, String, String)
      */
-    public FeedbackSessionResultsBundle getFeedbackSessionResultsForInstructorInSection(String feedbackSessionName,
-            String courseId, String userEmail, String section, SectionDetail sectionDetail)
-            throws EntityDoesNotExistException {
-
+    public SessionResultsBundle getSessionResultsForUser(
+            String feedbackSessionName, String courseId, String userEmail, UserRole role,
+            @Nullable String questionId, @Nullable String section) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/api/Logic.java:1443]: We don't normally put `@Nullable` annotation at this layer. What's the advantage here?

File: src/main/java/teammates/ui/webapi/action/ActionFactory.java [hunk 4]
```diff
@@ -63,7 +63,6 @@
         map(ResourceURIs.RESPONSE_COMMENT, PUT, UpdateFeedbackResponseCommentAction.class);
         map(ResourceURIs.RESPONSE_COMMENT, DELETE, DeleteFeedbackResponseCommentAction.class);
         map(ResourceURIs.RESULT, GET, GetSessionResultsAction.class);
-        map(ResourceURIs.RESULT_CSV, GET, GetSessionResultsAsCsvAction.class);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/ActionFactory.java:66]: Why is this removed?

File: src/test/java/teammates/test/driver/GaeSimulation.java [hunk 5]
```diff
@@ -183,24 +182,6 @@ public void loginAsStudentInstructor(String userId) {
         assertFalse(user.isAdmin);
     }
 
-    /**
-     * Returns an {@link teammates.ui.controller.Action} object that matches the parameters given.
-     *
-     * @param parameters Parameters that appear in a HttpServletRequest received by the app.
-     */
-    public teammates.ui.controller.Action getLegacyActionObject(String uri, String... parameters) {
-        InvocationContext ic = invokeWebRequest(uri, parameters);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/GaeSimulation.java:192]: You can remove this method and the field `sc` of this class as well. Also you can remove the `httpunit` test dependency in `build.gradle`.

File: src/main/java/teammates/logic/api/Logic.java (hunk 3 above)
Comment (xpdavid) [reply src/main/java/teammates/logic/api/Logic.java:1443]: More explicitly indicates that the variables can be null.

For consistency with the code base, I can also put them in the Javadoc.

File: src/test/java/teammates/test/cases/automated/FeedbackSessionUpdateRespondentWorkerActionTest.java [hunk 6]
```diff
@@ -103,22 +101,32 @@ public void allTests() throws EntityDoesNotExistException {
...
-                fsLogic.getFeedbackSessionResponseStatus(session.getFeedbackSessionName(), session.getCourseId());
-        assertFalse(responseStatus.getStudentsWhoDidNotRespond().contains(respondentEmail));
-        if (!isInstructor) {
-            assertTrue(responseStatus.getStudentsWhoResponded().contains(respondentEmail));
+            boolean isInstructor) {
+        FeedbackSessionAttributes sessionInDb =
+                logic.getFeedbackSession(session.getFeedbackSessionName(), session.getCourseId());
+
+        Set<String> respondingStudentList = sessionInDb.getRespondingStudentList();
+        Set<String> respondingInstructorList = sessionInDb.getRespondingInstructorList();
```
[the remark this answers is not in the document]
Comment (Derek-Hardy) [reply src/test/java/teammates/test/cases/automated/FeedbackSessionUpdateRespondentWorkerActionTest.java:109]: Actually the method itself also named as `get...List` but in fact return a set.

File: src/test/java/teammates/test/cases/automated/FeedbackSessionUpdateRespondentWorkerActionTest.java [hunk 7]
```diff
@@ -103,22 +101,32 @@ public void allTests() throws EntityDoesNotExistException {
...
-                fsLogic.getFeedbackSessionResponseStatus(session.getFeedbackSessionName(), session.getCourseId());
-        assertTrue(responseStatus.getStudentsWhoDidNotRespond().contains(respondentEmail));
-        if (!isInstructor) {
-            assertFalse(responseStatus.getStudentsWhoResponded().contains(respondentEmail));
+            boolean isInstructor) {
+        FeedbackSessionAttributes sessionInDb =
+                logic.getFeedbackSession(session.getFeedbackSessionName(), session.getCourseId());
+
+        Set<String> respondingStudentList = sessionInDb.getRespondingStudentList();
+        Set<String> respondingInstructorList = sessionInDb.getRespondingInstructorList();
```
Comment (Derek-Hardy) [inline src/test/java/teammates/test/cases/automated/FeedbackSessionUpdateRespondentWorkerActionTest.java:125]: Similarly here

File: src/test/java/teammates/test/cases/datatransfer/CourseRosterTest.java [hunk 8]
```diff
@@ -104,15 +105,96 @@ public void testGetEmailToNameTableFromRoster() {
...
     private List<StudentAttributes> createStudentList(String... studentData) {
         List<StudentAttributes> students = new ArrayList<>();
         for (int i = 0; i < studentData.length; i += 2) {
             StudentAttributes student = StudentAttributes
-                    .builder("", "")
-                    .withName("")
+                    .builder("", studentData[i + 1])
+                    .withName(studentData[i])
+                    .withTeamName(studentData[i])
+                    .withSectionName(studentData[i] + "'s Section")
```
Comment (Derek-Hardy) [inline src/test/java/teammates/test/cases/datatransfer/CourseRosterTest.java:196]: Although it is just a test class, I think it is still good to name each element in the pair within the for loop for better understanding.
Like `studentTeam = studentData[i]`, `studentEmail = studentData[i + 1]`.

File: src/main/java/teammates/ui/webapi/output/SessionResultsData.java [hunk 9]
```diff
@@ -371,18 +437,24 @@ public CommentOutput getParticipantComment() {
...
-        public CommentOutput(FeedbackResponseCommentAttributes frc, String commentGiverName, String lastEditorName) {
+        public CommentOutput(FeedbackResponseCommentAttributes frc,
+                             String commentGiver, String commentGiverName,
+                             String lastEditorEmail, String lastEditorName) {
             super(frc);
+            this.commentGiver = commentGiver;
+            this.commentGiverName = commentGiverName;
+            this.lastEditorEmail = lastEditorEmail;
+            this.lastEditorName = lastEditorName;
+        }
```
Comment (Derek-Hardy) [inline src/main/java/teammates/ui/webapi/output/SessionResultsData.java:449]: So, decided to pass in `null` and remove if-else check? :)

File: src/test/java/teammates/test/cases/automated/FeedbackSessionUpdateRespondentWorkerActionTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (xpdavid) [reply src/test/java/teammates/test/cases/automated/FeedbackSessionUpdateRespondentWorkerActionTest.java:109]: The method names are for legacy reason. Let not refactor them in this PR.

File: src/test/java/teammates/test/cases/datatransfer/CourseRosterTest.java (hunk 8 above)
Comment (xpdavid) [reply src/test/java/teammates/test/cases/datatransfer/CourseRosterTest.java:196]: The names are implicitly specified by the builder pattern?

File: src/main/java/teammates/ui/webapi/action/ActionFactory.java (hunk 4 above)
Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/ActionFactory.java:66]: Here are my investigations after trying to bring back the function:

I think we only have two options here because of the couplings:

**Keep the second commit**:
- Need to rewrite the CSV generation in the backend.
- Or write a new function in the frontend for CSV generation.
 
Personally, I think the CSV generation should be in the frontend where we care about the data presentation (Actually the current refactor is targeted to make this happened in the frontend to minimized the code duplication (e.g., missing response generation)). 

I have done some initial attempts and I would say it requires the same efforts to do either. After all, the CSV view is just a non-interactive version of the per question view in the frontend.

**Remove the second commit**:
- This will always work as commits are independent.
- We will remove other functions later. Not sure when though.

Personally, I would like to be ambitious and do it in frontend with the interns :P

File: src/test/java/teammates/test/cases/datatransfer/CourseRosterTest.java (hunk 8 above)
Comment (Derek-Hardy) [reply src/test/java/teammates/test/cases/datatransfer/CourseRosterTest.java:196]: There's no indication what `studentData[i + 1]` is when instantiating the builder.

File: src/main/java/teammates/ui/webapi/output/SessionResultsData.java (hunk 9 above)
Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/output/SessionResultsData.java:560]: Applied builder pattern instead 👍

Comment (Derek-Hardy) [reply src/main/java/teammates/ui/webapi/output/SessionResultsData.java:560]: Looks great!

File: src/main/java/teammates/common/datatransfer/CourseRoster.java [hunk 10]
```diff
@@ -104,4 +118,76 @@ private void populateInstructorListByEmail(List<InstructorAttributes> instructor
...
+     * Gets info of a participant associated with an identifier in the course.
+     *
+     * @return an object {@link ParticipantInfo} containing the name, teamName and the sectionName.
+     */
+    public ParticipantInfo getInfoForIdentifier(String identifier) {
+        String name = Const.USER_NOBODY_TEXT;
+        String teamName = Const.USER_NOBODY_TEXT;
+        String sectionName = Const.DEFAULT_SECTION;
+
+        boolean isStudent = this.getStudentForEmail(identifier) != null;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/CourseRoster.java:140]: Nitpicking, but this is Java so `this` is not exactly necessary.
